src/main/java/com/upserve/uppend/BlockedLongs.java

```diff
@@ -223,39 +212,90 @@ public LongStream values(Long pos) {
             return LongStream.empty();
         }
 
+        long[] longs = valuesArray(pos);
+        return Arrays.stream(longs);
+    }
+
+    public long[] valuesArray(Long pos) {
+
         if (pos < 0 || pos > size()) {
             log.error("Bad position value {} in file {} of size {}", pos, file, size());
+            return new long[]{};
+        }
+
+        // size | -next
+        // prev | -last
+        final long size = readLong(pos);
+
+        if (size < 0) {
+            long nextPos = -size;
+            long[] values = new long[valuesPerBlock];
+            for (int i = 0; i < valuesPerBlock; i++) {
+                values[i] = readLong(pos + 16 + i * 8);
+            }
+
+            long[] additionalValues = valuesArray(nextPos);
+
+            if (additionalValues.length == 0){
+                return values;
+            } else {
+                long[] result = new long[valuesPerBlock + additionalValues.length];
+                System.arraycopy(values,0,result,0, valuesPerBlock);
+                System.arraycopy(additionalValues,0,result,valuesPerBlock, additionalValues.length);
+                return result;
+            }
+        } else if (size > valuesPerBlock) {
+            throw new IllegalStateException("too high num values: expected <= " + valuesPerBlock + ", got " + size);
+        } else if (size == 0) {
+            return new long[]{};
+        } else {
+            int numValues = (int) size;
+            long[] values = new long[numValues];
+            for (int i = 0; i < numValues; i++) {
+                values[i] = readLong(pos + 16 + i * 8);
+            }
+            return values;
+        }
+    }
+
+    // Lazy values is much slower in Performance tests with a large number of blocks.
+    // Might be workable with a custom spliterator that was block aware for parallelization
+    public LongStream lazyValues(Long pos) {
+        log.trace("streaming values from {} at {}", file, pos);
+
+        valuesReadCounter.increment();
+
+        if (pos == null) {
+            // pos will be null for missing keys
             return LongStream.empty();
         }
 
-        ByteBuffer buf = readBlock(pos);
-        if (buf == null) {
+        if (pos < 0 || pos > size()) {
+            log.error("Bad position value {} in file {} of size {}", pos, file, size());
             return LongStream.empty();
         }
-        buf.flip();
 
         // size | -next
         // prev | -last
-
-        long size = buf.getLong();
-        buf.getLong();
+        final long size = readLong(pos);
 
         if (size < 0) {
             long nextPos = -size;
             long[] values = new long[valuesPerBlock];
             for (int i = 0; i < valuesPerBlock; i++) {
-                values[i] = buf.getLong();
+                values[i] = readLong(pos + 16 + i * 8);
             }
             return LongStreams.lazyConcat(Arrays.stream(values), () -> values(nextPos));
         } else if (size > valuesPerBlock) {
```

Comment:
Reorganize with `if(size > valuesPerBlock)` and `if(size == 0)` at top?

Reply:
Commented out the lazy method entirely.